### PR TITLE
T22-swap indices

### PR DESCRIPTION
**Pull request by Harshini-Jesi**

### CODE REVIEW

File: Program.cs [hunk 1]
```diff
@@ -1,7 +1,16 @@
...
-            Console.WriteLine ("Hello, World!");
-        }
-    }
-}
\ No newline at end of file
+﻿int number = GetInput ("Enter a number : ");
+char[] value = Convert.ToString (number).ToCharArray ();
+Console.WriteLine ("Enter the index positions to be swapped:");
+int a = GetInput ("FromPosition = "), b = GetInput ("ToPosition = ");
+swap (value, ref a, ref b);
```
Comment (varalakshmi-imm) [inline Program.cs:5]: Is ref required for this?

File: Program.cs [hunk 2]
```diff
@@ -1,7 +1,16 @@
...
+
+int GetInput (string prompt) {
+   while (true) {
+      Console.Write (prompt);
+      if (int.TryParse (Console.ReadLine (), out int value)) return value;
+      else continue;
+   }
+}
+
+void swap (char[] value, ref int a, ref int b) => (value[b], value[a]) = (value[a], value[b]);
```
Comment (varalakshmi-imm) [inline Program.cs:16]: Arrays are of reference types. Check [this ](https://stackoverflow.com/questions/967402/are-arrays-or-lists-passed-by-default-by-reference-in-c)for more info.
image

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: Program.cs [hunk 3]
```diff
@@ -1,7 +1,20 @@
...
-        }
-    }
-}
\ No newline at end of file
+﻿int number = GetInput ("Enter a number : ");
+char[] value = Convert.ToString (number).ToCharArray ();
+while (true) {
+   Console.WriteLine ("Enter the index positions to be swapped:");
+   int a = GetInput ("FromPosition = "), b = GetInput ("ToPosition = ");
+   if (a <= value.Length && b <= value.Length) {
```
Comment (varalakshmi-imm) [inline Program.cs:6]: Not handling -ve values. If -1 is given the program crashes.

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]
You can also enable XML documentation for the project which enables xml type comments as that is the practice expected.
For eg:
image

You should enable _Documentation file_ option in Build/Output in the properties of the project as shown below:
image

File: Program.cs [hunk 4]
```diff
@@ -1,7 +1,27 @@
...
+      }
+      break;
+   } else continue;
+}
+
+int GetInput (string prompt) {
+   while (true) {
+      Console.Write (prompt);
+      if (int.TryParse (Console.ReadLine (), out int value) && value >= 0) return value;
+      else continue;
```
Comment (varalakshmi-imm) [inline Program.cs:23]: You are returning if (condition == true) so else is not required in the next line. Just continue as if condition fails naturally the next statement will get executed.

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: Training.csproj [hunk 5]
```diff
@@ -1,9 +1,11 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>net7.0</TargetFramework>
     <ImplicitUsings>enable</ImplicitUsings>
+    <GenerateDocumentationFile>True</GenerateDocumentationFile>
+    <DocumentationFile>H:\harshini.s\learn\t22.txt</DocumentationFile>
```
Comment (varalakshmi-imm) [inline Training.csproj:8]: image
I am not able to compile it so leave it blank for the XML file path. Let it use the default path. Also there are warnings which have to be fixed. For XML comments, there should be class.

File: .editorconfig.txt [hunk 6]
```diff
@@ -0,0 +1,40 @@
+[user]
```
Comment (varalakshmi-imm) [inline .editorconfig.txt:1]: image
Why is this .editorconfig.txt added?

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: .editorconfig.txt (hunk 6 above)
Comment (Harshini-Jesi) [reply .editorconfig.txt:1]: I have changed to Trumpf laptop, network and mail id. So, mail id and paths of notepad, winmerge and P4merge has been changed in .editorconfig file.

Comment (varalakshmi-imm) [reply .editorconfig.txt:1]: Fine. But only one .editorconfig is required. 
If .editorconfg.txt is the new one then rename it to .editorconfig.

File: Program.cs [hunk 7]
```diff
@@ -1,7 +1,25 @@
...
+      int number = GetInput ("Enter a positive number : ");
+      char[] arr = Convert.ToString (number).ToCharArray ();
+      int max = arr.Length - 1;
+      Console.WriteLine ($"Enter the index positions to be swapped within 0 to {max}:");
+      int from = GetInput ("FromPosition = ", max);
+      int to = GetInput ("ToPosition = ", max);
+      Swap (arr, from, to);
+      Console.WriteLine ($"Index swapped number: {new string (arr)}");
+   }
+   ///<summary>Gets the input from the user until it is a positive number</summary>
```
Comment (varalakshmi-imm) [inline Program.cs:12]: Leave a blank line before this line. In general there should be a blank line between methods or properties.

File: Program.cs [hunk 8]
```diff
@@ -1,7 +1,25 @@
...
+   ///<param name="max">if set to -1 no range check is done, otherwise checks if value is less than or equal to max</param>
+   static int GetInput (string prompt, int max = -1) {
+      while (true) {
+         Console.Write (prompt);
+         if (!int.TryParse (Console.ReadLine (), out int value) || value < 0) continue;
+         if (max == -1 || value <= max) return value;
+      }
+   }
+
+   static void Swap (char[] arr, int from, int to) => (arr[to], arr[from]) = (arr[from], arr[to]);
```
Comment (varalakshmi-imm) [inline Program.cs:23]: Comments missing for this method

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]
Comments at the program or class level is missing.
Write what this program is doing in a couple of lines.

File: Program.cs [hunk 9]
```diff
@@ -1,7 +1,32 @@
...
+      int max = arr.Length - 1;
+      Console.WriteLine ($"Enter the index positions to be swapped within 0 to {max}:");
+      int from = GetInput ("FromPosition = ", max);
+      int to = GetInput ("ToPosition = ", max);
+      Swap (arr, from, to);
+      Console.WriteLine ($"Index swapped number: {new string (arr)}");
+   }
+
+   ///<summary>Gets the input from the user until it is a positive number</summary>
+   ///<param name="prompt">Displays the prompt for the user to enter the input</param>
```
Comment (varalakshmi-imm) [inline Program.cs:16]:
```suggestion
   /// <param name="prompt">Displays the prompt for the user to enter the input</param>

```
@Harshini-Jesi Always leave a space after /// or //

**Review by varalakshmi-imm — APPROVED** [review]

**Pull request closed without merging**